### ISSUE TITLE
Exact spawns scarcity refactor

### ISSUE THREAD

**Issue by caligari87** [caligari87/Ugly-as-Sin#31]
As-is, scarcity rolls randomly per spawn to determine removal. While on average this is fine, the RNG can *in theory* remove or keep all spawns even at the lowest setting, and players have complained of this.

This change would make scarcity more exact, by pre-counting all the items of a type (via thinkeriterator) and then removing a set percentage of them randomly. 

Pseudocode:
```csharp
while (thing = ThingIterator.Next()) {
    array.Push(thing);
}

removeAmount = array.size() * percentage;

while(array.size() >  removeAmount) {
    index = random(0, Array.Size());
    mo = array[index];
    if(mo) {
        mo.destroy();
        array[index].Delete();
    }
}

array.Clear(); // just to be safe
```

**Comment by caligari87** [comment]
Initial work on [`exact_scarcity`](https://github.com/caligari87/Ugly-as-Sin/tree/exact_scarcity) branch. Testing required, but scarcity slider should be more consistent now. Overspawn still works as normal, with some minor tweaks to individual items.

**Comment by caligari87** [comment]
Merged into master with commit e446ad527f2a40e21a2194d522b03d4964cd40e5, seems to be functional as intended. Closing.
